src/expr/src/expr/mod.rs

```diff
@@ -79,13 +79,14 @@ pub use self::expr_unary::new_unary_expr;
 use super::Result;
 
 /// Instance of an expression
+#[async_trait::async_trait]
 pub trait Expression: std::fmt::Debug + Sync + Send {
     /// Get the return data type.
     fn return_type(&self) -> DataType;
 
     /// Eval the result with extra checks.
-    fn eval_checked(&self, input: &DataChunk) -> Result<ArrayRef> {
-        let res = self.eval(input)?;
+    async fn eval_checked(&self, input: &DataChunk) -> Result<ArrayRef> {
```

Comment:
just curious, is this zero-cost in the case where the function does not contain any actually async blocks that depend on I/O?

Reply:
I was also thinking about this problem. Guess it's not 🤪

Reply:
btw I found this article. might be interesting read
https://thenewstack.io/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/

Reply:
It's NOT. In fact every async function call in the `async-trait` creates a boxed future, which introduces dynamic memory allocation. It won't be a big problem if it is not called frequently, for example in batch evaluation. But I can't tell how much it will impact on evaluation by row. I'll do a bench for this before merging.